File: jetson_infer/common/network/MQTTClient.cpp
#include "MQTTClient.h"
#include <iostream>

MQTTClient::MQTTClient(const std::string& address, int port, const std::string& clientId, const std::string& username, const std::string& password)
        : address_(address), clientId_(clientId), port_(port), isConnected_(false) {

    mosquitto_lib_init();
    mosq_ = mosquitto_new(clientId_.c_str(), true, this);
    if (!mosq_) {
        throw std::runtime_error("Failed to initialize Mosquitto client.");
    }

    if (!username.empty() && !password.empty()) {
        mosquitto_username_pw_set(mosq_, username.c_str(), password.c_str());
    }

    mosquitto_connect_callback_set(mosq_, MQTTClient::on_connect);
    mosquitto_disconnect_callback_set(mosq_, MQTTClient::on_disconnect);
    mosquitto_message_callback_set(mosq_, MQTTClient::on_message);
}

MQTTClient::~MQTTClient() {
    disconnect();
    mosquitto_destroy(mosq_);
    mosquitto_lib_cleanup();
}

void MQTTClient::setMessageCallback(std::function<void(const std::string&, const void*, size_t)> callback) {
    messageCallback_ = std::move(callback);
}

bool MQTTClient::connect() {
    int ret = mosquitto_connect(mosq_, address_.c_str(), port_, 60);
    return checkError(ret, "connect");
}

void MQTTClient::disconnect() {
    if (mosq_ && isConnected_) {
        int ret = mosquitto_disconnect(mosq_);
        if (checkError(ret, "disconnect")) {
            isConnected_ = false;  // 标记为未连接
        }
    }
}

bool MQTTClient::publish(const std::string& topic, const void* payload, size_t payloadlen) {
    int ret = mosquitto_publish(mosq_, nullptr, topic.c_str(), payloadlen, payload, 0, false);
    return checkError(ret, "publish");
}

bool MQTTClient::subscribe(const std::string& topic) {
    int ret = mosquitto_subscribe(mosq_, nullptr, topic.c_str(), 0);
    return checkError(ret, "subscribe");
}

// 手动处理事件循环
bool MQTTClient::listen(int timeout) {
    int ret = mosquitto_loop(mosq_, timeout, 1);  // 使用非阻塞模式
    return checkError(ret, "listen");
}

// Static callback functions
void MQTTClient::on_message(struct mosquitto* mosq, void* obj, const struct mosquitto_message* msg) {
    auto* client = static_cast<MQTTClient*>(obj);
    if (client->messageCallback_) {
        // 传递主题、payload（未转换为字符串）和长度
        client->messageCallback_(msg->topic, msg->payload, msg->payloadlen);
    }
}

void MQTTClient::on_connect(struct mosquitto* mosq, void* obj, int rc) {
    if (rc == 0) {
        std::cout << "Connected to the MQTT broker successfully." << std::endl;
        auto* client = static_cast<MQTTClient*>(obj);
        client->isConnected_ = true;  // 设置连接成功
    } else {
        std::cerr << "Failed to connect: " << mosquitto_strerror(rc) << std::endl;
    }
}

void MQTTClient::on_disconnect(struct mosquitto* mosq, void* obj, int rc) {
    auto* client = static_cast<MQTTClient*>(obj);
    if (rc == 0) {
        std::cout << "Disconnected from the MQTT broker gracefully." << std::endl;
    } else {
        std::cerr << "Unexpected disconnection from the MQTT broker: " << mosquitto_strerror(rc) << std::endl;
    }
    client->isConnected_ = false;
}

// 统一错误检查和日志输出
bool MQTTClient::checkError(int resultCode, const std::string& operation) {
    if (resultCode != MOSQ_ERR_SUCCESS) {
        std::cerr << "Failed to " << operation << ": " << mosquitto_strerror(resultCode) << std::endl;
        return false;
    }
    return true;
}

File: obsoleted/displayer/common/mqtt/MQTTClient.py
import paho.mqtt.client as mqtt
from typing import Callable, Optional, Any


class MQTTClient:
    def __init__(self, address: str, port: int, client_id: str, username: str = "", password: str = ""):
        self.address = address
        self.port = port
        self.client_id = client_id
        self.username = username
        self.password = password
        self.client = mqtt.Client(client_id=client_id)
        self.is_connected = False
        self.message_callback: Optional[Callable[[str, bytes, Any], None]] = None
        self.message_callback_arg: Any = None

        if username and password:
            self.client.username_pw_set(username, password)

        self.client.on_connect = self._on_connect
        self.client.on_disconnect = self._on_disconnect
        self.client.on_message = self._on_message

    def set_message_callback(self, callback: Callable[[str, bytes, Any], None], arg: Any = None):
        self.message_callback = callback
        self.message_callback_arg = arg

    def connect(self) -> bool:
        try:
            self.client.connect(self.address, self.port, 60)
            return True
        except Exception as e:
            print(f"Failed to connect: {e}")
            return False

    def disconnect(self):
        if self.is_connected:
            self.client.disconnect()

    def publish(self, topic: str, payload: bytes) -> bool:
        try:
            self.client.publish(topic, payload)
            return True
        except Exception as e:
            print(f"Failed to publish: {e}")
            return False

    def subscribe(self, topic: str) -> bool:
        try:
            self.client.subscribe(topic)
            return True
        except Exception as e:
            print(f"Failed to subscribe: {e}")
            return False

    def listen(self, timeout: float = 1.0) -> bool:
        try:
            self.client.loop(timeout=timeout)
            return True
        except Exception as e:
            print(f"Error in message loop: {e}")
            return False

    def _on_connect(self, client, userdata, flags, rc):
        if rc == 0:
            print("Connected to the MQTT broker successfully.")
            self.is_connected = True
        else:
            print(f"Failed to connect: {mqtt.connack_string(rc)}")

    def _on_disconnect(self, client, userdata, rc):
        if rc == 0:
            print("Disconnected from the MQTT broker gracefully.")
        else:
            print(f"Unexpected disconnection from the MQTT broker: {mqtt.error_string(rc)}")
        self.is_connected = False

    def _on_message(self, client, userdata, msg):
        if self.message_callback:
            self.message_callback(msg.topic, msg.payload, self.message_callback_arg)

    def __del__(self):
        self.disconnect()